File: backend/app/tools/bookings.py
from __future__ import annotations

from typing import Any

from app.models import ToolResult
from app.tools.common import citation_for_csv, infer_room, read_csv


def lookup(payload: dict[str, Any]) -> ToolResult:
    message = payload["message"]
    sender = message.get("sender", "").lower()
    room = infer_room(message)
    matches: list[tuple[int, dict[str, str]]] = []
    for line_no, row in read_csv("bookings.csv"):
        if row["email"].lower() == sender or (room and row["room"] == room):
            matches.append((line_no, row))
    matches.sort(key=lambda item: (item[1]["email"].lower() != sender, item[1]["status"] != "checked_in"))
    booking = matches[0][1] if matches else {}
    citations = [citation_for_csv("bookings.csv", line_no, row, "guest_name") for line_no, row in matches[:2]]
    return ToolResult(tool="bookings.lookup", data={"booking": booking, "room": room}, citations=citations)

File: pibackend/src/tools/bookings.ts
import { readCsv } from "../lib/csv";
import type { ToolPayload, ToolResult } from "../types";
import { citationForCsv, inferRoom } from "./common";

export function lookup(payload: ToolPayload): ToolResult {
  const sender = payload.message.sender.toLowerCase();
  const room = inferRoom(payload.message);
  const matches = readCsv("bookings.csv").filter(
    ([, row]) => row.email.toLowerCase() === sender || Boolean(room && row.room === room)
  );
  matches.sort((a, b) => {
    const aSender = a[1].email.toLowerCase() !== sender ? 1 : 0;
    const bSender = b[1].email.toLowerCase() !== sender ? 1 : 0;
    const aStatus = a[1].status !== "checked_in" ? 1 : 0;
    const bStatus = b[1].status !== "checked_in" ? 1 : 0;
    return aSender - bSender || aStatus - bStatus;
  });
  return {
    tool: "bookings.lookup",
    data: { booking: matches[0]?.[1] ?? {}, room },
    citations: matches.slice(0, 2).map(([lineNo, row]) => citationForCsv("bookings.csv", lineNo, row, "guest_name"))
  };
}
